distributed-commandbus-springcloud/src/main/java/org/axonframework/springcloud/commandhandling/SpringCloudHttpBackupCommandRouter.java

```diff
@@ -0,0 +1,141 @@
+                                              RestTemplate restTemplate) {
+        this(discoveryClient, routingStrategy, ACCEPT_ALL_INSTANCES_FILTER, restTemplate);
+    }
+
+    /**
+     * Initialize a {@link org.axonframework.commandhandling.distributed.CommandRouter} with the given {@link
+     * org.springframework.cloud.client.discovery.DiscoveryClient} to update its own membership as a {@code
+     * CommandRouter} and to create its own awareness of available nodes to send commands to in a {@link
+     * org.axonframework.commandhandling.distributed.ConsistentHash}.
+     * The {@code routingStrategy} is used to define the key based on which Command Messages are routed to their
+     * respective handler nodes.
+     * A {@code Predicate<ServiceInstance>} to filter a {@link org.springframework.cloud.client.ServiceInstance} from
+     * the membership update loop.
+     * The {@link org.springframework.web.client.RestTemplate} is used as a backup mechanism to request another member's
+     * {@link org.axonframework.springcloud.commandhandling.MessageRoutingInformation} with.
+     *
+     * @param discoveryClient       The {@code DiscoveryClient} used to discovery and notify other nodes
+     * @param routingStrategy       The strategy for routing Commands to a Node
+     * @param serviceInstanceFilter The {@code Predicate<ServiceInstance>} used to filter
+     * @param restTemplate          The {@code RestTemplate} used to request another member's {@link
+     *                              org.axonframework.springcloud.commandhandling.MessageRoutingInformation} with.
+     */
+    public SpringCloudHttpBackupCommandRouter(DiscoveryClient discoveryClient,
+                                              RoutingStrategy routingStrategy,
+                                              Predicate<ServiceInstance> serviceInstanceFilter,
+                                              RestTemplate restTemplate) {
+        super(discoveryClient, routingStrategy, serviceInstanceFilter);
+        this.restTemplate = restTemplate;
+        this.messageRoutingInfo = null;
+    }
+
+    @Override
+    public void updateMembership(int loadFactor, Predicate<? super CommandMessage<?>> commandFilter) {
+        messageRoutingInfo = new MessageRoutingInformation(loadFactor, commandFilter, serializer);
+        super.updateMembership(loadFactor, commandFilter);
+    }
+
+    @GetMapping
+    public MessageRoutingInformation getLocalMessageRoutingInformation() {
+        return messageRoutingInfo;
+    }
+
+    @Override
+    protected MessageRoutingInformation messageRoutingInformationFromNonMetadataSource(
+            ServiceInstance serviceInstance) {
+        SimpleMember<URI> simpleMember = buildSimpleMember(serviceInstance);
+        if (simpleMember.local()) {
+            return getLocalMessageRoutingInformation();
+        }
+
+        URI endpoint = simpleMember.getConnectionEndpoint(URI.class)
```

Comment:
This logic isn't really "falling back". It should first attempt to retrieve MessageRoutingInformation based on the ServiceInstance information retrieved from the Discovery mechanism. Only if that can't be found for the specific node, then it should fall back to doing an HTTP call.
It might be better to return an `Optional<MessageRoutingInformation>` here, indicating that it is very well possible that no information can be found. Non-axon nodes, for example. If that is the case, these nodes will (even after an http callback) return no information. The router should also "blacklist" these nodes somehow to make sure they are not interrogated for routing information on each 'heart beat'.

Reply:
For the first bit of your comment about 'not falling back':

This is an override of the `messageRoutingInformationFromNonMetadataSource()`, which is called in the `SpringCloudCommandRouter#getMessageRoutingInformationFrom(ServiceInstance)` function, only if the `ServiceInstance` the request is done for does not contain any of the required metadata fields.

Thus, there are two `messageRoutingInformationFrom...()` function at this point, the `messageRoutingInformationFromMetadata()` and the `messageRoutingInformationFromNonMetadataSource()` (which is overridden here).
The first is always performed firs, thus making this function the _fallback function_.

Reply:
Now the second part of your comment about `Optional` and black listing:

The `Optional` bit is implemented simple enough and I agree that's a good approach.
Hence the aforementioned function will return an `Optional<MessageRoutingInformation>` after I've done the implementation of that.

The black listing however deserves some more discussing.
I think we need to decide when we want to black list a node. On it's first exception? Or on the second? Or only on certain exception? Should that be configurable? To much options at this point, so lets decide on this some where soon.